src/trace/log.zig

```diff
@@ -171,79 +169,55 @@ pub const ChannelPrintLogger = struct {
-            return Entry{ .channel_print = ChannelPrintEntry.init(self.log_allocator, maybe_scope, self.channel, Level.debug) };
-        }
-        return .noop;
-    }
-
-    pub fn trace(self: *Self, comptime maybe_scope: ?[]const u8) Entry {
-        if (@intFromEnum(self.max_level) >= @intFromEnum(Level.trace)) {
-            return Entry{ .channel_print = ChannelPrintEntry.init(self.log_allocator, maybe_scope, self.channel, Level.trace) };
-        }
-        return .noop;
-    }
+    pub fn log(
+        self: *Self,
+        comptime scope: ?[]const u8,
+        level: Level,
+        fields: anytype,
+        comptime fmt: []const u8,
+        args: anytype,
+    ) void {
+        if (@intFromEnum(self.max_level) < @intFromEnum(level)) return;
+        const size = logfmt.countLog(scope, level, fields, fmt, args);
+        const msg_buf = self.allocBuf(size) catch |err| {
+            std.debug.print("allocBuff failed with err: {any}", .{err});
+            return;
+        };
+        var stream = std.io.fixedBufferStream(msg_buf);
+        logfmt.writeLog(stream.writer(), scope, level, fields, fmt, args) catch |err| {
+            std.debug.print("writeLog failed with err: {any}", .{err});
+            self.log_allocator.free(msg_buf);
+            return;
+        };
+        std.debug.assert(size == stream.pos);
 
-    pub fn log(self: *Self, comptime scope: ?[]const u8, level: Level, comptime message: []const u8) void {
-        if (@intFromEnum(self.max_level) >= @intFromEnum(level)) {
-            switch (level) {
-                .err => self.err(scope).log(message),
-                .warn => self.warn(scope).log(message),
-                .info => self.info(scope).log(message),
-                .debug => self.debug(scope).log(message),
-                .trace => self.trace(scope).log(message),
-            }
-        }
+        self.channel.send(msg_buf) catch |err| {
+            std.debug.print("Send msg through channel failed with err: {any}", .{err});
+            self.log_allocator.free(msg_buf);
+            return;
+        };
     }
 
-    pub fn logf(self: *Self, comptime scope: ?[]const u8, level: Level, comptime fmt: []const u8, args: anytype) void {
-        if (@intFromEnum(self.max_level) >= @intFromEnum(level)) {
-            switch (level) {
-                .err => self.err(scope).logf(fmt, args),
-                .warn => self.warn(scope).logf(fmt, args),
-                .info => self.info(scope).logf(fmt, args),
-                .debug => self.debug(scope).logf(fmt, args),
-                .trace => self.trace(scope).logf(fmt, args),
-            }
+    // Utility function for allocating memory from RecycleFBA for part of the log message.
+    fn allocBuf(self: *const Self, size: u64) ![]u8 {
+        for (0..100) |_| {
+            return self.log_allocator.alloc(u8, size) catch {
+                std.time.sleep(std.time.ns_per_ms);
+                if (self.exit.load(.unordered)) {
```

Comment:
Needs to be `monotonic`.

Reply:
d9dc8e71c73207e9d2d5ea03b3c8878a32be553a